core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js

```diff
@@ -0,0 +1,165 @@
+
+/**
+ * @fileoverview Service which handles opening and closing
+ * the training data of an answer group.
+ */
+
+oppia.factory('TrainingDataEditorPanelService', [
+  '$rootScope', '$uibModal', 'AlertsService', 'UrlInterpolationService',
+  function($rootScope, $uibModal, AlertsService, UrlInterpolationService) {
+    return {
+      openTrainingDataEditor: function(externalSave) {
+        AlertsService.clearWarnings();
+        if (externalSave) {
+          $rootScope.$broadcast('externalSave');
+        }
+        $uibModal.open({
+          templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+            '/pages/exploration_editor/editor_tab/' +
+            'training_data_editor_directive.html'),
+          backdrop: true,
+          controller: [
+            '$scope', '$injector', '$uibModalInstance',
+            'ExplorationStatesService', 'EditorStateService',
+            'AnswerClassificationService', 'ExplorationContextService',
+            'stateInteractionIdService', 'AngularNameService',
+            'EXPLICIT_CLASSIFICATION', 'TRAINING_DATA_CLASSIFICATION',
+            'ExplorationHtmlFormatterService', 'ResponsesService',
+            'stateCustomizationArgsService', 'TrainingDataService',
+            'TrainingModalService',
+            function($scope, $injector, $uibModalInstance,
+                ExplorationStatesService, EditorStateService,
+                AnswerClassificationService, ExplorationContextService,
+                stateInteractionIdService, AngularNameService,
+                EXPLICIT_CLASSIFICATION, TRAINING_DATA_CLASSIFICATION,
+                ExplorationHtmlFormatterService, ResponsesService,
+                stateCustomizationArgsService, TrainingDataService,
+                TrainingModalService) {
+              var _explorationId = ExplorationContextService.getExplorationId();
+              var _stateName = EditorStateService.getActiveStateName();
+              var _state = ExplorationStatesService.getState(_stateName);
+              var answerGroupIndex = (
+                ResponsesService.getActiveAnswerGroupIndex());
+
+              $scope.stateContent = _state.content.getHtml();
+              $scope.inputTemplate = (
+                ExplorationHtmlFormatterService.getInteractionHtml(
+                  stateInteractionIdService.savedMemento,
+                  stateCustomizationArgsService.savedMemento,
+                  false, 'testInteractionInput'));
+              $scope.trainingData = [];
+
+              var _rebuildTrainingData = function() {
+                $scope.trainingData = [];
+                TrainingDataService.getTrainingDataOfAnswerGroup(
+                  answerGroupIndex).forEach(function(answer) {
+                  answerTemplate = (
+                    ExplorationHtmlFormatterService.getAnswerHtml(
+                      answer, stateInteractionIdService.savedMemento,
+                      stateCustomizationArgsService.savedMemento));
```

Comment:
Just to check, what happens to the training data if the customization args change, say (e.g. if we delete/rearrange choices in MultipleChoiceInput)?

I guess that might not be an issue because we don't need ML for MultipleChoiceInput, but I couldn't find the place where this PR depends on feconf.INTERACTION_CLASSIFIER_MAPPING.

Reply:
That case has not been handled in the current infrastructure. Also this PR does not depend on `feconf.INTERACTION_CLASSIFIER_MAPPING` because that mapping used when we train a classifier on training data. But training data will there only when interaction is trainable. So if some interaction is not trainable then no part of this interface will be enabled. So no training data modification will be allowed in that case.

Reply:
Ah, OK, that's interesting.

Can we make the presence of all this stuff depend on feconf.INTERACTION_CLASSIFIER_MAPPING (in addition to whether there are answers to train)? I'm thinking about the case where we expose a trainable interaction that we later decide is non-trainable, and it seems like a good idea to me to have the additional check as a safeguard (defense-in-depth).

Reply:
The interaction classifier mapping is separate from training interface. Its purpose is to provide info whether a classifier on Oppia-ml is available to train the training data of given interaction. I think that case is entirely separate from whether interaction should have training data. Also, the mapping is used between Oppia and Oppia-ml and is not transferred to the frontend. I think the check for isInteractionTrainable() is sufficient.

Additionally, it keeps things loosely coupled. We can always make an interaction non-trainable even if we have a classifier on Oppia-ml (like if the classifier is bad or something). Why do you suggest we should have deeper checks?

Reply:
Perhaps because I missed the check for isInteractionTrainable() -- sorry! Using that would also be fine IMO. Just some way to ensure that this isn't shown for all interactions.

That said, I can't seem to find isInteractionTrainable() when doing Ctrl+F on this page, except for code which has been deleted. Could you please point me to its use?

Reply:
The name is `isCurrentInteractionTrainable`. My mistake. Also, take a look at this [line](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/exploration_editor/editor_tab/state_editor_responses.html#L98) in state_editor_responses.html.

So, `Teach Oppia` button and ``Modify Training Data` both are guarded using `isCurrentInteractionTrainable()`.

Reply:
Excellent, thanks! Sounds good then.